book/src/development/example_early.md

```diff
@@ -0,0 +1,395 @@
+# Example with `EarlyLintPass`
+
+Let us dig into an actual Clippy lint that implements `EarlyLintPass`
+so that we can put the knowledge from all previous chapters into practice.
+
+## The Problem: `pub use ...`
+
+The usage of `pub use ...` is extremely common in Rust projects.
+However, for certain projects we might want to restrict writing
+`pub use ...` to prevent unintentional exports or to encourage
+the developers to place exported items explicitly in public modules.
+
+For instance, we might often write the following module export:
+
+```rust
+pub mod outer {
+    mod inner {
+        pub struct Test {}
+    }
+    pub use inner::Test;  // <- `pub use` instance
+}
+
+use outer::Test;
+```
+
+Instead, for some very specific projects, we want to encourage
+the following approach instead:
+
+```rust
+pub mod outer {
+    pub struct Test {}
+}
+
+use outer::Test;
+```
+
+## Define the Lint
+
+Since we only need to detect the usage of `pub use` and do not need
+any type information, we could implement `EarlyLintPass` for this lint.
+
+Let us name it `pub_use`, which suggests explicitly what this lint
+is aimed to do.
+
+Additionally, since restricting `pub use ...` is for very specific cases,
+we probably want to create a lint but put it into `clippy::restriction`
```

Comment:
I wouldn't chose an restriction lint here. Those lints are for really special use cases only and people might not relate to it and won't see the motivation. There is probably a simple early lint that is warn-by-default?